lib/scholar/neighbors/utils.ex

```diff
@@ -2,11 +2,89 @@ defmodule Scholar.Neighbors.Utils do
   @moduledoc false
   import Nx.Defn
   require Nx
+  import Scholar.Shared
+
+  defn linear_search(data, query, opts) do
+    k = opts[:num_neighbors]
+    metric = opts[:metric]
+    batch_size = opts[:batch_size]
+
+    type = Nx.Type.merge(to_float_type(data), to_float_type(query))
+
+    distance_fn =
+      case metric do
+        {:minkowski, p} ->
+          fn data, query ->
+            query
+            |> Nx.new_axis(1)
+            |> Nx.subtract(data)
+            |> Nx.pow(p)
+            |> Nx.sum(axes: [2])
+          end
+
+        {:cosine} ->
+          &Scholar.Metrics.Distance.pairwise_cosine/2
+      end
+
+    {query_size, dim} = Nx.shape(query)
+    num_batches = div(query_size, batch_size)
+    leftover_size = rem(query_size, batch_size)
+
```

Comment:
This part might also need to be moved inside `Scholar.Shared` as a function that takes a tensor and batch size and returns `{batches, leftover}` where `batches` is a tensor of shape `{num_batches, batch_size, dim}` and `leftover` is a tensor of shape `{leftover_size, dim}`.
Might be relevant for #246, task 1.

Reply:
Yes, definitely. In Python, there is even a function divmod that do that